src/site/content/en/blog/form-fatigue-validate-user-input/index.md

```diff
@@ -0,0 +1,308 @@
+    from entering invalid dates. For example, prevent users from selecting a
+    delivery date in the past or a delivery time outside of your operating hours:
+
+<figure class="w-figure">
+  <img src="image-browser-date-selector.png" alt="browser built-in date selector" style="max-width: 400px;">
+  <figcaption class="w-figcaption">
+    Chrome's default date picker.
+  </figcaption>
+</figure>
+
+
+```html
+<input type="date" min="2019-03-25" max="2019-04-25">  
+<input type="time" min="11:00" max="17:00">  
+```
+
+{% Aside 'caution' %}
+Browsers that do not support `date` or `time` fallback to type `text`. Always manually validate input on form submission.
+{% endAside %}
+
+## Use the `pattern` attribute
+
+In addition to restricting input, you can validate input against a regular
+expression using the `pattern` attribute. For example, a valid US zip code is 5
+digits and sometimes followed by a dash and an additional 4 digits:
+
+
+```html  
+<input type="text" pattern="^\d{5,6}(?:[-\s]\d{4})?$" ...>  
+```
+
+The following regular expression validates credit card numbers:
+
+
+```html
+^(?:4[0-9]{12}(?:[0-9]{3})?|5[1-5][0-9]{14}|3[47][0-9]{13}|3(?:0[0-5]|[68][0-9])[0-9]{11}|6(?:011|5[0-9]{2})[0-9]{12}|(?:2131|1800|35\d{3})\d{11})$  
+```
+
+{% Aside 'note' %}
+Unlike `pattern`, `min`, `max`, `minlength` and `maxlength` will restrict user input in real-time.
+{% endAside %}
+
+## Custom validation
+
+If built-in validation attributes aren't enough, use the
+[Constraint Validation API](https://w3c.github.io/html/sec-forms.html#constraints)
+to manually mark fields invalid. To manually invalidate a field, use
+`setCustomValidity()` to pass a custom validation message to the invalid input
+element. 
+
+For example, you can immediately validate a shipping address using a back-end location service:
+
+```js/3
+let shippingAddress = shippingAddressInput.value;  
+if (!checkValidAddress(shippingAddress)) {  
+  // invalid card number, set custom validity message  
+  shippingAddressInput.setCustomValidity('This is not a valid address.');  
+} else {  
+  // valid card number, reset message  
+  shippingAddressInput.setCustomValidity('');  
+}  
+```  
+Be sure to reset the error message when the field becomes valid:  
+
+<!-- TODO (mdiblasio): how to bold in templates?  -->
+
+```js/6
+let shippingAddress = shippingAddressInput.value;  
+if (!checkValidAddress(shippingAddress)) {  
+  // invalid card number, set custom validity message  
+  shippingAddressInput.setCustomValidity('This is not a valid address.');  
+} else {  
+  // valid card number, reset message  
+  shippingAddressInput.setCustomValidity('');  
+}  
+``` 
+
+### Set custom validation messages
+
+Default validation messages can be vague and difficult to understand. Use the [Constraint Validation API](https://w3c.github.io/html/sec-forms.html#constraints) to provide helpful, specific and non-technical validation messages when users enter invalid information.
```

Comment:
This paragraph almost reads like the one above it so I wasn't sure what was different. It feels a bit odd to introduce the Constraint Validation API again here. Maybe it could just continue on from the section above?

Reply:
consolidated